public/app/features/alerting/unified/components/rules/central-state-history/EventListSceneObject.tsx

```diff
@@ -85,12 +86,23 @@ interface HistoryLogEventsProps {
   logRecords: LogRecord[];
 }
 function HistoryLogEvents({ logRecords }: HistoryLogEventsProps) {
+  const { page, pageItems, numberOfPages, onPageChange } = usePagination(logRecords, 1, PAGE_SIZE);
   return (
-    <ul>
-      {logRecords.map((record) => {
-        return <EventRow key={record.timestamp + (record.line.fingerprint ?? '')} record={record} />;
-      })}
-    </ul>
+    <Stack direction="column" gap={0}>
+      <ul>
+        {pageItems.map((record) => {
+          return (
+            <EventRow
+              key={record.timestamp + (record.line.fingerprint ?? '')}
+              record={record}
+              logRecords={logRecords}
+            />
+          );
+        })}
+      </ul>
+      {/* This paginations improves the performance considerably , making the page load faster */}
+      <Pagination currentPage={page} numberOfPages={numberOfPages} onNavigate={onPageChange} hideWhenSinglePage />
```

Comment:
nit: can we also add a `@todo` here that we might want to switch to infinite scrolling at some point?

Reply:
I opted for pagination over infinite scroll because pagination allows users to navigate directly to the last page (some users have lots of events, and events are sorted by time, so pagination can be more useful)